source/distributed/solution_transfer.cc

```diff
@@ -207,26 +229,97 @@ namespace parallel
     SolutionTransfer<dim, VectorType, DoFHandlerType>::pack_callback(
       const typename Triangulation<dim, DoFHandlerType::space_dimension>::
         cell_iterator &cell_,
-      const typename Triangulation<dim, DoFHandlerType::space_dimension>::
-        CellStatus /*status*/)
+      const typename Triangulation<dim,
+                                   DoFHandlerType::space_dimension>::CellStatus
+        status)
     {
       typename DoFHandlerType::cell_iterator cell(*cell_, dof_handler);
 
       // create buffer for each individual object
       std::vector<::dealii::Vector<typename VectorType::value_type>> dofvalues(
         input_vectors.size());
 
-      auto cit_input = input_vectors.cbegin();
-      auto it_output = dofvalues.begin();
-      for (; cit_input != input_vectors.cend(); ++cit_input, ++it_output)
+      if (has_hp_dof_handler)
         {
-          it_output->reinit(cell->get_fe().dofs_per_cell);
-          cell->get_interpolated_dof_values(*(*cit_input), *it_output);
+          unsigned int dofs_per_cell = numbers::invalid_unsigned_int;
+          unsigned int fe_index      = numbers::invalid_unsigned_int;
+
+          switch (status)
+            {
+              case parallel::distributed::Triangulation<
+                dim,
+                DoFHandlerType::space_dimension>::CELL_PERSIST:
+              case parallel::distributed::Triangulation<
+                dim,
+                DoFHandlerType::space_dimension>::CELL_REFINE:
+                dofs_per_cell = cell->get_fe().dofs_per_cell;
+                fe_index      = cell->active_fe_index();
+                break;
+
+              case parallel::distributed::Triangulation<
+                dim,
+                DoFHandlerType::space_dimension>::CELL_COARSEN:
+                // In case of coarsening, we need to find a suitable fe index
+                // for the parent cell. We choose the 'least face dominating
+                // fe index' on all children from the associated FECollection.
+                {
+                  const hp::DoFHandler<dim, DoFHandlerType::space_dimension>
+                    *hp_dof_handler = dynamic_cast<
+                      const hp::DoFHandler<dim, DoFHandlerType::space_dimension>
+                        *>(&*dof_handler);
```

Comment:
why do you need this cast? shouldn't the cast be the identity here? in other words, shouldn't `DoFHandlerType` and `hp::DoFHandler<dim, DoFHandlerType::space_dimension>` be the same type?

Reply:
`dof_handler->get_fe()` returns a `FiniteElement` object in the non-hp case, but a `FECollection` in the hp case, whose member functions we require (i.e. `find_least_face_dominating_fe` and `operator[]`).

So I could either cast the `dof_handler` like now, or the returned object of `get_fe()` to a `FECollection`. Or do you have a different idea?

Reply:
You can use ‘get_fe_collection’ which returns the same regardless of hp/non-hp

Reply:
`hp::DofHandler::get_fe()` is deprecated, and the other access functions are consistent.
`get_fe(int i)` returns a FiniteElement and `get_fe_collection()` returns a `FECollection`.

Reply:
Thank you! I'll give it a try!